UnnecessaryParenthesesCheck does not flag unnecessary parentheses in conditional expression

I have read check documentation: https://checkstyle.org/checks/coding/unnecessaryparentheses.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```
PS D:\CS\test> javac src/Test.java
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <property name="haltOnException" value="true"/>
  <module name="TreeWalker">
    <module name="UnnecessaryParentheses">
    </module>
  </module>
</module>
PS D:\CS\test> cat src/Test.java

public class Test {
    void test(int x) {
        boolean a = (x == 3); // violation
        boolean b = (!(x == 3)); // violation

        boolean c = (x == 3) ? (x == 3) : (!(x == 3));  // expected 3 violations
    }

}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml  src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:4:21: Unnecessary parentheses around assignment right-hand side. [UnnecessaryParentheses]
[ERROR] D:\CS\test\src\Test.java:5:21: Unnecessary parentheses around assignment right-hand side. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 2 errors.


```
---

**Describe what you expect in detail.**

`UnnecessaryParentheses` should flag unnecessary parentheses in conditional expressions. We should add `Question` to the acceptable tokens or something.


---

--- a/config/checkstyle-checks.xml
+++ b/config/checkstyle-checks.xml
@@ -576,7 +576,8 @@
                        BSR_ASSIGN, BXOR_ASSIGN, DIV_ASSIGN, MINUS_ASSIGN, MOD_ASSIGN, PLUS_ASSIGN,
                        SL_ASSIGN, SR_ASSIGN, STAR_ASSIGN, LAMBDA, TEXT_BLOCK_LITERAL_BEGIN, LAND,
                        LOR, LITERAL_INSTANCEOF, GT, LT, GE, LE, EQUAL, NOT_EQUAL, UNARY_MINUS,
-                       UNARY_PLUS, INC, DEC, LNOT, BNOT, POST_INC, POST_DEC, BOR, BXOR, BAND"/>
+                       UNARY_PLUS, INC, DEC, LNOT, BNOT, POST_INC, POST_DEC, BOR, BXOR, BAND,
+                       QUESTION"/>
     </module>
     <module name="UnnecessarySemicolonAfterOuterTypeDeclaration"/>
     <module name="UnnecessarySemicolonAfterTypeMemberDeclaration"/>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java
@@ -19,6 +19,9 @@
 
 package com.puppycrawl.tools.checkstyle.checks.coding;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.regex.Pattern;
 
 import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
@@ -441,6 +444,7 @@ public class UnnecessaryParenthesesCheck extends AbstractCheck {
             TokenTypes.BXOR,
             TokenTypes.BOR,
             TokenTypes.BAND,
+            TokenTypes.QUESTION,
         };
     }
 
@@ -457,6 +461,10 @@ public class UnnecessaryParenthesesCheck extends AbstractCheck {
 
         if (isLambdaSingleParameterSurrounded(ast)) {
             log(ast, MSG_LAMBDA);
+        }
+        else if (ast.getType() == TokenTypes.QUESTION) {
+            getParenthesesChildrenAroundQuestion(ast)
+                .forEach(unnecessaryChild -> log(unnecessaryChild, MSG_EXPR));
         }
         else if (parent.getType() != TokenTypes.ANNOTATION_MEMBER_VALUE_PAIR) {
             final int type = ast.getType();
@@ -682,6 +690,27 @@ public class UnnecessaryParenthesesCheck extends AbstractCheck {
             }
         }
         return result;
+    }
+
+    /**
+     *  Returns the direct LPAREN tokens children to a given QUESTION token which
+     *  contain an expression not a literal variable.
+     *
+     *  @param questionToken {@code DetailAST} question token to be checked
+     *  @return the direct children to the given question token which their types are LPAREN
+     *          tokens and not contain a literal inside the parentheses
+     */
+    private static List<DetailAST> getParenthesesChildrenAroundQuestion(DetailAST questionToken) {
+        final List<DetailAST> surroundedChildren = new ArrayList<>();
+        DetailAST directChild = questionToken.getFirstChild();
+        while (directChild != null) {
+            if (directChild.getType() == TokenTypes.LPAREN
+                    && !TokenUtil.isOfType(directChild.getNextSibling(), LITERALS)) {
+                surroundedChildren.add(directChild);
+            }
+            directChild = directChild.getNextSibling();
+        }
+        return Collections.unmodifiableList(surroundedChildren);
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
@@ -357,4 +357,22 @@ public class UnnecessaryParenthesesCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputUnnecessaryParenthesesWhenExpressions.java"), expected);
     }
+
+    @Test
+    public void testUnnecessaryParenthesesConditionalExpression() throws Exception {
+        final String[] expected = {
+            "19:17: " + getCheckMessage(MSG_EXPR),
+            "19:29: " + getCheckMessage(MSG_LITERAL, "3"),
+            "19:35: " + getCheckMessage(MSG_LITERAL, "4"),
+            "25:18: " + getCheckMessage(MSG_EXPR),
+            "28:18: " + getCheckMessage(MSG_EXPR),
+            "28:33: " + getCheckMessage(MSG_EXPR),
+            "35:26: " + getCheckMessage(MSG_EXPR),
+            "36:17: " + getCheckMessage(MSG_EXPR),
+            "36:41: " + getCheckMessage(MSG_EXPR),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputUnnecessaryParenthesesConditionalExpression.java"), expected);
+
+    }
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesConditionalExpression.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesConditionalExpression.java
@@ -1,0 +1,42 @@
+/*
+UnnecessaryParentheses
+tokens = EXPR, IDENT, NUM_DOUBLE, NUM_FLOAT, NUM_INT, NUM_LONG, \
+         STRING_LITERAL, LITERAL_NULL, LITERAL_FALSE, LITERAL_TRUE, ASSIGN, \
+         BAND_ASSIGN, BOR_ASSIGN, BSR_ASSIGN, BXOR_ASSIGN, DIV_ASSIGN, \
+         MINUS_ASSIGN, MOD_ASSIGN, PLUS_ASSIGN, SL_ASSIGN, SR_ASSIGN, STAR_ASSIGN, \
+         LAMBDA, TEXT_BLOCK_LITERAL_BEGIN, LAND, LITERAL_INSTANCEOF, GT, LT, GE, \
+         LE, EQUAL, NOT_EQUAL, UNARY_MINUS, UNARY_PLUS, INC, DEC, LNOT, BNOT, \
+         POST_INC, POST_DEC, QUESTION
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
+
+public class InputUnnecessaryParenthesesConditionalExpression {
+    void method() {
+        int w = 5;
+        int x = (w == 3) ? (3) : (4);
+        // 3 violations above
+        //  'Unnecessary parentheses around expression'
+        //  'Unnecessary parentheses around literal '3''.
+        //  'Unnecessary parentheses around literal '4''.
+        int y = !(w>x) ? 3 : 4;
+        int z1 = (!(y >= w)) ? 5 : 6; // violation 'Unnecessary parentheses around expression'
+
+        int z2 = 5 > 3 ? 2 : 1;
+        int z3 = (z2 > 0) ? 5 : (z2 < -10) ? 7 : 3;
+        // 2 violations above
+        //  'Unnecessary parentheses around expression'
+        //  'Unnecessary parentheses around expression'
+
+        Object o = "";
+        // violation below 'Unnecessary parentheses around expression'
+        boolean result = (o instanceof String) ?
+                (o instanceof String) : (!(o instanceof String));
+        // 2 violations above
+        //  'Unnecessary parentheses around expression'
+        //  'Unnecessary parentheses around expression'
+
+    }
+}

--- a/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/Example3.java
+++ b/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/Example3.java
@@ -1,0 +1,24 @@
+/*xml
+<module name="Checker">
+  <module name="TreeWalker">
+    <module name="UnnecessaryParentheses">
+      <property name="tokens" value="QUESTION" />
+    </module>
+  </module>
+</module>
+*/
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
+
+// xdoc section -- start
+class Example3 {
+
+  void method() {
+    int a = 9, b = 8;
+    int c = (a > b) ? 1 : 0; // violation 'Unnecessary parentheses around expression'
+
+    int d = c == 1 ? (b % 2 == 0) ? 1 : 0 : 5;
+    // violation above 'Unnecessary parentheses around expression'
+  }
+
+}
+// xdoc section -- end

--- a/src/xdocs/checks/coding/unnecessaryparentheses.xml
+++ b/src/xdocs/checks/coding/unnecessaryparentheses.xml
@@ -177,6 +177,8 @@ if ((++f) &gt; g &amp;&amp; a) { // violation, unnecessary paren
                     BOR</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#BAND">
                     BAND</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#QUESTION">
+                    QUESTION</a>
                   .
               </td>
               <td>
@@ -372,6 +374,33 @@ class Example2 {
     if (x &gt;= 0 &amp; (x &lt;= 8 ^ y &lt;= 11) &amp; y &gt;= 8) {
       return;
     }
+  }
+
+}
+        </source>
+        <p id="Example3-config">
+          To configure the check to detect unnecessary parentheses around conditional expressions
+          <code> '?' </code>:
+        </p>
+        <source>
+&lt;module name=&quot;Checker&quot;&gt;
+  &lt;module name=&quot;TreeWalker&quot;&gt;
+    &lt;module name=&quot;UnnecessaryParentheses&quot;&gt;
+      &lt;property name=&quot;tokens&quot; value=&quot;QUESTION&quot; /&gt;
+    &lt;/module&gt;
+  &lt;/module&gt;
+&lt;/module&gt;
+        </source>
+        <p id="Example3-code">Example:</p>
+        <source>
+class Example3 {
+
+  void method() {
+    int a = 9, b = 8;
+    int c = (a &gt; b) ? 1 : 0; // violation 'Unnecessary parentheses around expression'
+
+    int d = c == 1 ? (b % 2 == 0) ? 1 : 0 : 5;
+    // violation above 'Unnecessary parentheses around expression'
   }
 
 }

--- a/src/xdocs/checks/coding/unnecessaryparentheses.xml.template
+++ b/src/xdocs/checks/coding/unnecessaryparentheses.xml.template
@@ -117,6 +117,21 @@ if ((++f) &gt; g &amp;&amp; a) { // violation, unnecessary paren
                  value="resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/Example2.java"/>
           <param name="type" value="code"/>
         </macro>
+        <p id="Example3-config">
+          To configure the check to detect unnecessary parentheses around conditional expressions
+          <code> '?' </code>:
+        </p>
+        <macro name="example">
+          <param name="path"
+                 value="resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/Example3.java"/>
+          <param name="type" value="config"/>
+        </macro>
+        <p id="Example3-code">Example:</p>
+        <macro name="example">
+          <param name="path"
+                 value="resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/Example3.java"/>
+          <param name="type" value="code"/>
+        </macro>
       </subsection>
 
       <subsection name="Example of Usage" id="Example_of_Usage">

